deep_qa/models/sentence_selection/siamese_sentence_selector.py

```diff
@@ -0,0 +1,123 @@
+from typing import Any, Dict
+from overrides import overrides
+from keras.layers import Input
+
+from ...data.instances.sentence_selection_instance import SentenceSelectionInstance
+from ...layers.attention.attention import Attention
+from ...layers.wrappers.encoder_wrapper import EncoderWrapper
+from ...training.text_trainer import TextTrainer
+from ...training.models import DeepQaModel
+
+
+class SiameseSentenceSelector(TextTrainer):
+    """
+    This class implements a (generally) Siamese network for the answer
+    sentence selectiont ask. Given a question and a collection of sentences,
+    we aim to identify which sentence has the answer to the question. This
+    model encodes the question and each sentence with (possibly different)
+    encoders, and then does a cosine similarity and normalizes to get a
+    distribution over the set of sentences.
+
+    Note that in some cases, this may not be exactly "Siamese" because the
+    question and sentences encoders can differ.
+    """
+    def __init__(self, params: Dict[str, Any]):
+        self.num_question_words = params.pop('num_question_words', None)
+        self.num_sentence_words = params.pop('num_sentence_words', None)
+        self.num_sentences = params.pop('num_sentences', None)
+        super(SiameseSentenceSelector, self).__init__(params)
+
+    @overrides
+    def _build_model(self):
+        """
+        The basic outline here is that we'll pass the questions and each
+        sentence in the passage through some sort of encoder (e.g. BOW, GRU,
+        or biGRU).
+
+        Then, we take the encoded representation of the question and calculate
+        a cosine similarity with the encoded representation of each sentence in
+        the passage, to get a tensor of cosine similarities with shape
+        (batch_size, num_sentences_per_passage). We then normalize for each
+        batch to get a probability distribution over sentences in the passage.
+        """
+        # First we create input layers and pass the inputs through embedding layers.
+        # shape: (batch size, question_length)
+        question_input = Input(shape=self._get_sentence_shape(self.num_question_words),
+                               dtype='int32', name="question_input")
+
+        # shape: (batch size, num_sentences, sentences_length)
+        sentences_input_shape = ((self.num_sentences,) +
+                                 self._get_sentence_shape(self.num_sentence_words))
```

Comment:
No need to have the argument to `self._get_sentence_shape()`, because the default is to use `self.num_sentence_words`.  Also, the shapes in the comments should use names that match the variables.